src/Symfony/Component/Security/Core/Authentication/Token/NullToken.php

```diff
@@ -28,9 +28,9 @@ public function getRoleNames(): array
         return [];
     }
 
-    public function getUser()
+    public function getUser(): UserInterface
     {
-        return '';
+        throw new \BadMethodCallException('Cannot get user on a NullToken.');
```

Comment:
not sure this is the correct approach here

Reply:
#42631 should make this PR's CI green

Reply:
This means we're making the `NullToken` implementation part of the abstract token API. Aka the API requires checking `instanceof NullToken` is userland. This is not a minor consideration. Is that really what we want?

Reply:
Don't we need a NullUser instead?

Reply:
I think you're right. Userland voters are likely to call `getUser()` blindly, which would break with this exception.
Adding NullUser on 5.4 seems to be the best option

Reply:
For reference, this discussion continued in https://github.com/symfony/symfony/pull/42644#issuecomment-901803728